packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart

```diff
@@ -0,0 +1,92 @@
+import 'package:flutter_test/flutter_test.dart';
+import 'package:mockito/mockito.dart';
+import 'package:os_detect/override.dart';
+import 'package:shared_preferences/shared_preferences.dart';
+
+import 'amplify_push_noitfications_impl_test.mocks.dart';
+import 'test_data/fake_notification_messges.dart';
+
+void testGlobalCallbackFunction(PushNotificationMessage pushMessage) {
+  expect(
+    pushMessage.title,
+    PushNotificationMessage.fromJson(standardAndroidPushMessage).title,
+  );
+}
+
+void main() {
+  late AmplifyPushNotificationsFlutterApi flutterApi;
+
+  setUp(() {
+    TestWidgetsFlutterBinding.ensureInitialized();
+    flutterApi = AmplifyPushNotificationsFlutterApi.instance;
+  });
+
+  test('should queue events when the serviceClient is not ready', () {
+    SharedPreferences.setMockInitialValues({});
+    flutterApi.onNotificationReceivedInBackground(standardAndroidPushMessage);
+    expect(flutterApi.eventQueue.length, 1);
+  });
+
+  test('should flush events when the serviceClient is ready', () async {
+    SharedPreferences.setMockInitialValues({});
+    await flutterApi
+        .onNotificationReceivedInBackground(standardAndroidPushMessage);
+    expect(flutterApi.eventQueue.length, 2);
+    final mockServiceClient = MockServiceProviderClient();
+    when(
+      mockServiceClient.recordNotificationEvent(
+        eventType: anyNamed('eventType'),
+        notification: anyNamed('notification'),
+      ),
+    ).thenAnswer(
+      (_) async {},
+    );
+    flutterApi.serviceProviderClient = mockServiceClient;
+    await Future.delayed(const Duration(seconds: 3), () {});
+    expect(flutterApi.eventQueue.length, 0);
+  });
+
+  test('should call the external callback function', () async {
+    SharedPreferences.setMockInitialValues({});
+    void externalCallback(PushNotificationMessage pushMessage) {
+      expect(
+        pushMessage.title,
+        PushNotificationMessage.fromJson(standardAndroidPushMessage).title,
+      );
+    }
+
+    flutterApi.registerOnReceivedInBackgroundCallback(externalCallback);
+    await flutterApi
+        .onNotificationReceivedInBackground(standardAndroidPushMessage);
+  });
+
+  test('should invoke the top-level or static external callback function',
+      () async {
+    await overrideOperatingSystem(
+      const OperatingSystem('android', ''),
+      () async {
+        SharedPreferences.setMockInitialValues({});
+        final pref = await SharedPreferences.getInstance();
+        AmplifyPushNotifications(
+          serviceProviderClient: MockServiceProviderClient(),
+          backgroundProcessor: () async => {},
+          // dependencyManager: dependencyManager,
+        ).onNotificationReceivedInBackground(
+          testGlobalCallbackFunction,
+        );
+        flutterApi
+            .registerOnReceivedInBackgroundCallback(testGlobalCallbackFunction);
+
+        await Future.delayed(const Duration(seconds: 3), () {});
```

Comment:
It's just an arbitrary time I set to wait for the callback function to be registered. Since we want to keep the `onNotificationReceivedInBackground` async we are not waiting for the registration process to finish, hence I wait before triggering the function that is responsible to handle the notification coming from the native side.

Reply:
If it's arbitrary please use a smaller delay to make the tests run faster. The tests may be roughly 6 seconds slower waiting for this arbitrary time :D

Reply:
If you implement the `expectAsync` pattern with a completer, you can avoid this wait

Reply:
I am having to wait in-between callback registration and notification arrival because when registering we are using `unawaited` within the `onNotificationReceivedInBackground` [function](https://github.com/aws-amplify/amplify-flutter/blob/feat/push/android-utils/packages/notifications/push/amplify_push_notifications/lib/src/amplify_push_notifications_impl.dart#L162-L167). This choice was made to allow for using the API without awaiting it but as we can see it can introduce problems if the notification arrives too early as mimicked in our test by the use of `flutterApi.onNotificationReceivedInBackground`. 

I don't completely understand how completer is useful here but one proposal I have is to make the `onNotificationReceivedInBackground` API function async which will ensure the async operation to store the callback handle in shared preference is successful before it's ready to receive notifications.